### ISSUE TITLE
Implement `TArray`

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#1177]
Implement `zio.stm.TArray`:

```scala
class TArray[A] private (array: Array[TRef[A]]) { ... }
```

with methods similar to `Array` inside it, allowing access and updating by index, as well as folding and other types of functionality compatible with STM.

**Comment by LGLO** [comment]
I'll take it!

**Comment by LGLO** [comment]
@jdegoes 
I'm wondering about type of initial value for `fold`:
`def fold[E, E1 <: E , E2 <: E](acc: STM[E1, A], op: (A, A) => STM[E2, A]): STM[E, A]`
`def fold[E](acc: A)(op: (A, A) => STM[E, A]): STM[E, A]`

In case of former - should I add "syntax" class to enable two parameter lists syntax?

**Comment by jdegoes** [comment]
@LGLO 

There would be at least 2 versions on `TArray`:

Non-effectful:

```
def fold[Z](acc: Z)(op: (Z, A) => Z): STM[Nothing, Z]
```

Effectful:

```
def foldM[E, Z](acc: Z)(op: (Z, A) => STM[E, Z]): STM[E, Z]
```

Both of these should infer fine without syntax classes.

**Comment by ghostdogpr** [comment]
Closed by #1191
